### PR TITLE
Preprocessing, code generation minor fixes

### PR DESCRIPTION
**Pull request by tbennun**
- [x] Fix for repeated applications of closure resolver
- [x] Add `dace.nounroll` preprocessing hint
- [x] Better error printouts on missing arguments
- [x] Fix degenerate 1D copies of strided arrays
- [x] Enable pre-padding (start offset) for arrays
- [x] Use `json.dump` instead of `json.dumps` to save memory
- [x] Fix serialization of string dtype
- [x] Improve support for multiple callbacks with same name

### CODE REVIEW

File: dace/sdfg/state.py [hunk 1]
```diff
@@ -56,6 +56,17 @@ class StateGraphView(object):
     def __init__(self, *args, **kwargs):
         self._clear_scopedict_cache()
 
+    ###################################################################
+    # Typing overrides
+    
+    @overload
```
Comment (alexnick83) [inline dace/sdfg/state.py:62]: Is this to assist static type analysis (mypy, etc.)?

File: dace/codegen/targets/cpp.py [hunk 2]
```diff
@@ -420,15 +420,13 @@ def ndcopy_to_strided_copy(
             and ((src_copylen == copy_length and dst_copylen == copy_length) or
                  (tuple(src_shape) == tuple(copy_shape) and tuple(dst_shape) == tuple(copy_shape)))):
         # Emit 1D copy of the whole array
-        copy_shape = [functools.reduce(lambda x, y: x * y, copy_shape)]
-        return copy_shape, [1], [1]
+        return [src_copylen], [1], [1]
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/codegen/targets/cpp.py:423]: copy_shape is not equivalent to the actual length. The bug here was that nontrivial strides (padding) were not counted in the copy shape. In this case, `src_copylen` is equal to `dst_copylen`.

File: dace/data.py [hunk 3]
```diff
@@ -391,7 +391,9 @@ class Array(Data):
                          desc='This pointer may alias with other pointers in '
                          'the same function')
 
-    alignment = Property(dtype=int, default=0, desc='Allocation alignment in bytes (0 uses ' 'compiler-default)')
+    alignment = Property(dtype=int, default=0, desc='Allocation alignment in bytes (0 uses compiler-default)')
+
+    start_offset = Property(dtype=int, default=0, desc='Allocation offset elements for manual alignment (pre-padding)')
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/data.py:396]: `start_offset` is just padding, regardless of shape (for alignment), whereas `offset` is equivalent to what FORTRAN has as offset (i.e., does the first element start from 1, 0 or -50). We can document this.

File: dace/sdfg/state.py (hunk 1 above)
Comment (tbennun) [reply dace/sdfg/state.py:62]: Yes

File: dace/sdfg/utils.py [hunk 4]
```diff
@@ -1118,46 +1118,44 @@ def inline_sdfgs(sdfg: SDFG, permissive: bool = False, progress: bool = None, mu
...
     start = time.time()
 
-    for sd in reversed(sdfgs):
-        id = sd.sdfg_id
-        for state in sd.nodes():
-            for node in state.nodes():
-                if (progress is None and tqdm is not None and (time.time() - start) > 5):
-                    progress = True
-                    pbar = tqdm(total=len(sdfgs), desc='Inlining SDFG', initial=counter)
+    for ctr, (node, state) in enumerate(reversed(nsdfgs)):
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/sdfg/utils.py:1127]: the iterator change was only to make getting the "true" `ctr` easier. The progress report was not finishing at 100% but a bit before (due to SDFGs that were inlined/not-inlined before the progress printouts started). Now it always finishes at 100%

**Pull request merged**